src/xmerl_dsig.erl

```diff
@@ -0,0 +1,731 @@
+
+verify_invalid_test() ->
+    {Doc, _} = xmerl_scan:string("<x:foo xmlns:x=\"urn:foo:x:\"><x:name>blah</x:name></x:foo>", [{namespace_conformant, true}]),
+    {error, no_signature} = verify(Doc).
+
+verify_unknown_cert_test() ->
+    {Doc, _} = xmerl_scan:string("<?xml version=\"1.0\"?><x:foo ID=\"9616e6c0-f525-11b7-afb7-5cf9dd711ed3\" xmlns:x=\"urn:foo:x:\"><ds:Signature xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm=\"http://www.w3.org/2001/10/xml-exc-c14n#\"/><ds:SignatureMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#rsa-sha1\"/><ds:Reference URI=\"#9616e6c0-f525-11b7-afb7-5cf9dd711ed3\"><ds:Transforms><ds:Transform Algorithm=\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\"/></ds:Transforms><ds:DigestMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#sha1\"/><ds:DigestValue>V8qJXW/fet+RCUswqa+PSNLTG5Y=</ds:DigestValue></ds:Reference></ds:SignedInfo><ds:SignatureValue>B5rQVbuK6861DStrYldTEdM17wkec/adr08uAXWZXU7b3r+JKt0AcDRsZUgh3dsg</ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509SubjectName>O='Default Company Ltd';2.5.4.7='Default City';C=XX</ds:X509SubjectName><ds:X509Certificate>MIIBrTCCAWegAwIBAgIJAJsPdOI20ZF2MA0GCSqGSIb3DQEBBQUAMEIxCzAJBgNVBAYTAlhYMRUwEwYDVQQHDAxEZWZhdWx0IENpdHkxHDAaBgNVBAoME0RlZmF1bHQgQ29tcGFueSBMdGQwHhcNMTMwNTAyMDYwMDM0WhcNMjMwNTAyMDYwMDM0WjBCMQswCQYDVQQGEwJYWDEVMBMGA1UEBwwMRGVmYXVsdCBDaXR5MRwwGgYDVQQKDBNEZWZhdWx0IENvbXBhbnkgTHRkMEwwDQYJKoZIhvcNAQEBBQADOwAwOAIxAM0Wz0qz1bnRjfr5+lqs2HMk+MomI/qMy5SmjJ2HBH2OgZSqjKu3mg4tPzxjRG33mwIDAQABo1AwTjAdBgNVHQ4EFgQU2XTi/8L82oGx9mcaSMggervenTowHwYDVR0jBBgwFoAU2XTi/8L82oGx9mcaSMggervenTowDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQUFAAMxAELu647IINJuZT/vxZoEgBrAwQMK+l/yam5iAWQI5Y+NtCrbC16Vu0qkLSVP5Ednrw==</ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><x:name>blah</x:name></x:foo>", [{namespace_conformant, true}]),
+    {error, cert_not_accepted} = verify(Doc, [<<198>>]).
+
+verify_bad_digest_test() ->
+    {Doc, _} = xmerl_scan:string("<?xml version=\"1.0\"?><x:foo ID=\"9616e6c0-f525-11b7-afb7-5cf9dd711ed3\" xmlns:x=\"urn:foo:x:\"><ds:Signature xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm=\"http://www.w3.org/2001/10/xml-exc-c14n#\"/><ds:SignatureMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#rsa-sha1\"/><ds:Reference URI=\"#9616e6c0-f525-11b7-afb7-5cf9dd711ed3\"><ds:Transforms><ds:Transform Algorithm=\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\"/></ds:Transforms><ds:DigestMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#sha1\"/><ds:DigestValue>xPVYXCs5uMMmIbfTiTZ5R5DVhTU=</ds:DigestValue></ds:Reference></ds:SignedInfo><ds:SignatureValue></ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509Certificate></ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><x:name>b1ah</x:name></x:foo>", [{namespace_conformant, true}]),
+    {error, {bad_digest, _, _}} = verify(Doc).
+
+verify_bad_signature_test() ->
```

Comment:
Please add a comment about what is wrong with the signature.

Is it malformed or is it simply incorrect with regards to the document/digest?

Reply:
I don't understand. The test is called bad_digest, the xml has a bad_digest, and the code returns that it is a bad digest.
Do we need a comment saying that this xml has a bad digest?

If it was malformed it wouldn't return `bad_digests`, that is only returned if there is a bad_digest.

Reply:
The test I'm commenting on is called `verify_bad_signature_test`. If not, the PR UI is tricking me as that was the test I wanted to ask about.

Reply:
Ok. That is a test from the original author with a good digest, but a broken signature. So it will never verify.

Reply:
Ah, I realise now that `cert_not_accepted` covers my original question as well.